### PR TITLE
Move teams out of ddrace gamecontroller (#7127)

### PR DESCRIPTION
**Pull request by ChillerDragon**
Getting rid of some of the casts. Road to moddable ddnet with gamecontroller support.

## Checklist

- [ ] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [ ] Considered possible null pointers and out of bounds array indexing
- [ ] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

### CODE REVIEW

**Comment by heinrich5991** [comment]
Does this make sense? This implies other game modes also want to have DDRace

**Comment by ChillerDragon** [comment]
> Does this make sense? This implies other game modes also want to have DDRace teams.

Yea ddrace teams are nice aren't they? Also they are opt in anyways so who would not want them. What would be the alternative to get rid of the cast? If there is a technical solution for that then there is still a bunch of stuff referencing it. Such as chat commands and similar. Are you suggesting that should all be included in the controller? Or somehow made optional if the active controller is not a ddrace controller?

I don't get your point. Are you saying that everything that a ddnet based mod might not need should go into the ddrace controller? What about freeze? Draggers? Race finishes and database ranks?

I can not think of a single mod where teams in the gamecontroller hurt. If they have no use case in the mod then one can just set sv_teams 0 or comment out the /team chat command.

**Comment by heinrich5991** [comment]
Well, I don't really know where we want to go with the game controllers, I guess.

What's the purpose of having one at all, in the DDNet codebase? What purpose is it supposed to solve?

**Comment by ChillerDragon** [comment]
> Well, I don't really know where we want to go with the game controllers, I guess.

The casts currently make it impossible to create a new game controller. Like vanilla does it with ctf, dm and mods.

> What's the purpose of having one at all, in the DDNet codebase? What purpose is it supposed to solve?

The purpose is for forks to create their own gamecontroller. So they can modify ddnet without creating git conflicts everywhere.

**Comment by ChillerDragon** [comment]
> > > Well, I don't really know where we want to go with the game controllers, I guess.
> > 
> > 
> > The casts currently make it impossible to create a new game controller. Like vanilla does it with ctf, dm and mods.
> 
> But that doesn't mean that they should be moved upwards.

Disagree. But even if. Where else would they go? Stay in the ddrace controller? Then how to get rid of the casts?

**Comment by Robyt3** [comment]
> The casts currently make it impossible to create a new game controller. Like vanilla does it with ctf, dm and mods.

But this now assumes that every mod wants DDTeams, which is not the case for vanilla. Maybe we need something akin to [mixins](https://stackoverflow.com/a/18775236) for all optional gamecontroller components. The commands that use DDTeams should only be registered if the DDTeams component is used. `CGameContext` should not be aware of the DDTeams at all to get rid of most downcasts.

**Comment by Kaffeine** [comment]
We had discussed it for a bit at https://github.com/ddnet/ddnet/issues/7127

As I said at https://github.com/ddnet/ddnet/issues/7127#issuecomment-1712121609, now I think that it is better to have Teams be a part of the base game.

The questions to think about:
1. How hard it is to add teams if they are needed in the mod
2. How hard it is to disable the teams if they are not needed in the mod
3. How many mods need the teams

In Infclass we have "hook protection" feature which allows the player to control if teammates can hook them. At first I refuged the Teams and had some pains on removing it each time — it is not incapsulated into the DDRace GameController but instead it is used in few dozen classes — including CCharacter and CCharacterCore.

My points:
1. Having Teams constructed in DDRace GameController does not isolate it from the game and does not let other mods to have no Teams-related code. Instead, it forces the mods to either copy Teams initialization to their own ModGameController, or remove a lot of code causing a lot of merge conflicts. It makes no sense to keep Teams in the mod controller and have core classes depending on it.
2. It is relatively easy to make the Teams behave as the mod needs to — from vanilla behavior to something really custom (more complex than just co-existing player groups with disabled cross-interactions).

4.

File: src/game/server/gamecontext.cpp [hunk 1]
```diff
@@ -2033,7 +2032,7 @@ void CGameContext::OnSayNetMessage(const CNetMsg_Cl_Say *pMsg, int ClientID, con
 	if(Length == 0 || (pMsg->m_pMessage[0] != '/' && (g_Config.m_SvSpamprotection && pPlayer->m_LastChat && pPlayer->m_LastChat + Server()->TickSpeed() * ((31 + Length) / 32) > Server()->Tick())))
 		return;
 
-	int GameTeam = ((CGameControllerDDRace *)m_pController)->m_Teams.m_Core.Team(pPlayer->GetCID());
+	int GameTeam = (m_pController)->m_Teams.m_Core.Team(pPlayer->GetCID());
```
Comment (Kaffeine) [inline src/game/server/gamecontext.cpp:2035]: I'd suggest to use `CGameContext::GetDDRaceTeam()` API than spread this dependency on GameController, CTeams, CTeamsCore and it's API in those many places.

```suggestion
	int GameTeam = GetDDRaceTeam(pPlayer->GetCID());
```

File: src/game/server/gamecontext.cpp [hunk 2]
```diff
@@ -522,7 +522,7 @@ void CGameContext::SendChatTarget(int To, const char *pText, int Flags)
 void CGameContext::SendChatTeam(int Team, const char *pText)
 {
 	for(int i = 0; i < MAX_CLIENTS; i++)
-		if(((CGameControllerDDRace *)m_pController)->m_Teams.m_Core.Team(i) == Team)
+		if(m_pController->m_Teams.m_Core.Team(i) == Team)
```
Comment (Kaffeine) [inline src/game/server/gamecontext.cpp:525]: GetDDRaceTeam() (here and below)

**Comment by ChillerDragon** [comment]
> > The casts currently make it impossible to create a new game controller. Like vanilla does it with ctf, dm and mods.
> 
> But this now assumes that every mod wants DDTeams, which is not the case for vanilla

@Robyt3 ??? why does vanilla not want ddteams? Its nice to have multiple games at once on one server. Instead of spinning up 5 dm1 servers. And my point stands that this pr is not about what some mod might want or not want. Because things like freeze potentially are also not needed. How is that holding us back from getting rid of the casts?

**Comment by heinrich5991** [comment]
Thanks for chiming in, @Kaffeine. I agree with your reasoning.

@ChillerDragon Could you implement @Kaffeine's suggestions?

**Pull request closed without merging**